ibis/expr/operations/udf.py

```diff
@@ -0,0 +1,157 @@
+import itertools
+import typing
+from typing import TYPE_CHECKING, Any, Callable, TypeVar
+
+from public import public
+
+import ibis.common.exceptions as exc
+import ibis.expr.datatypes as dt
+import ibis.expr.operations as ops
+import ibis.expr.rules as rlz
+from ibis import util
+from ibis.common.annotations import attribute
+from ibis.common.collections import FrozenDict
+
+if TYPE_CHECKING:
+    import ibis.expr.types as ir
+
+
+@public
+class ScalarUDF(ops.Value):
+    @attribute.default
+    def output_shape(self):
+        return rlz.highest_precedence_shape(
+            itertools.chain.from_iterable(
+                util.promote_list(getattr(self, name)) for name in self.argnames
+            )
+        )
+
+
+@public
+class SingleScalarUDF(ScalarUDF):
+    pass
+
+
+@public
+class VectorizedScalarUDF(ScalarUDF):
+    pass
+
+
+def _wrap(wrapper, fn: Callable | None = None, *args: Any, **kwargs: Any) -> Callable:
+    """Wrap a function `fn` with `wrapper`, allowing zero arguments when used as part of a decorator."""
+    if fn is None:
+        return lambda fn: functools.update_wrapper(wrapper(fn, *args, **kwargs), fn)
+    return functools.update_wrapper(wrapper(fn, *args, **kwargs), fn)
+
+
+S1 = TypeVar("S1", bound=ScalarUDF)
+S2 = TypeVar("S2", bound=ScalarUDF)
+
+
+class UDF:
+    """Construct wrappers for user-defined functions."""
+
+    @util.experimental
+    def __call__(
+        self, fn: Callable | None = None, *args: Any, **kwargs: Any
+    ) -> Callable:
+        """Construct a scalar user-defined function.
+
+        Examples
+        --------
+        >>> from ibis.interactive import *
+        >>> @udf.scalar
+        ... def add_one(x: int) -> int:
+        ...     return x + 1
+        >>> @udf.scalar(schema="my_udfs")
+        ... def add_one(x: int) -> int:
+        ...     return x + 1
+        """
+        return _wrap(self._make_wrapper, fn, *args, **kwargs)
+
+    @util.experimental
+    def vectorized(
+        self, fn: Callable | None = None, *args: Any, **kwargs: Any
+    ) -> Callable:
+        """Construct a vectorized scalar user-defined function.
+
+        Examples
+        --------
+        >>> from ibis.interactive import *
+        >>> @udf.scalar.vectorized
+        ... def add_one(x: int) -> int:
+        ...     return x + 1
+        >>> @udf.scalar.vectorized(schema="my_udfs")
+        ... def add_one(x: int) -> int:
+        ...     return x + 1
+        """
+        return _wrap(self._make_vectorized_wrapper, fn, *args, **kwargs)
+
+    def _make_vectorized_wrapper(
+        self, fn: Callable, *args: Any, **kwargs: Any
+    ) -> Callable:
+        """Generate an ibis operation node type wrapper for a vectorized scalar function."""
+
+    def _make_wrapper(self, fn: Callable, *args: Any, **kwargs: Any) -> Callable:
+        """Generate an ibis operation node type wrapper for a simple scalar function."""
+
+    def make_node(self, fn: Callable, base: type[S1], *args, **kwargs) -> type[S2]:
```

Comment:
The previous udf implementation looks like the following:

```py
class VectorizedUDF(Value):
    func = rlz.instance_of((FunctionType, LambdaType))
    func_args = rlz.tuple_of(rlz.column(rlz.any))
    input_type = rlz.tuple_of(rlz.datatype)
    return_type = rlz.datatype

    @property
    def output_dtype(self):
        return self.return_type
```

What's the advantage of the following node factory compared to that? 
Does it do an extra layer of validation?

Reply:
> Does it do an extra layer of validation?

It's not an additional layer, but baked into how we're defining each node.

The inputs are not an arbitrary collection of expressions, they are built using rules based on the user's type annotations. This means the UDFs inputs are validated as much (or as little) as the user wants.